### PR TITLE
LibGfx/TIFF: Add support for tiled images :^)

### PR DESCRIPTION
**Pull request by LucasChollet**
Screenshot from 2024-02-09 21-09-36

---

It now even comes with a test :^)

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp [hunk 1]
```diff
@@ -92,7 +92,7 @@ class TIFFLoadingContext {
         if (segment_offsets()->size() != segment_byte_counts()->size())
             return Error::from_string_literal("TIFFImageDecoderPlugin: StripsOffset and StripByteCount have different sizes");
 
-        if (!m_metadata.rows_per_strip().has_value() && m_metadata.strip_byte_counts()->size() != 1)
+        if (!m_metadata.rows_per_strip().has_value() && segment_byte_counts()->size() != 1 && !m_metadata.tile_length().has_value())
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp:95]: nit: `!m_metadata.is_tiled()` would read nicer here (with the obvious impl)

File: Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp [hunk 2]
```diff
@@ -278,40 +278,45 @@ class TIFFLoadingContext {
...
                 Optional<Color> last_color {};
 
-                for (u32 column = 0; column < *m_metadata.image_width(); ++column) {
+                for (u32 column = 0; column < segment_width; ++column) {
+                    auto const absolute_column = column + segment_width * (segment_index % segment_per_rows);
+
                     if (metadata().photometric_interpretation() == PhotometricInterpretation::CMYK) {
                         auto const cmyk = TRY(read_color_cmyk(*decoded_stream));
-                        oriented_bitmap.get<ExifOrientedCMYKBitmap>().set_pixel(column, absolute_row, cmyk);
+                        if (absolute_column >= *m_metadata.image_width())
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp:317]: does this happen in practice? when? (maybe worth a comment?)

**Review by nico — APPROVED** [review]
Nice to see Cramps again!

File: Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp (hunk 2 above)
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp:317]: Yeah, quite often actually. Tiles are always padded, and this line allows us to skip this padding.

File: Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp [hunk 3]
```diff
@@ -284,16 +284,16 @@ class TIFFLoadingContext {
             auto decoded_stream = make<BigEndianInputBitStream>(move(decoded_strip));
 
             for (u32 row = 0; row < rows_per_strip; row++) {
-                auto const scanline = row + rows_per_strip * strip_index;
-                if (scanline >= *m_metadata.image_length())
+                auto const absolute_row = row + rows_per_strip * strip_index;
```
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp]: As in not relative to the begining of the segment. I added an explanation in the commit but it feels wrong to have to read the commit message to understand the name of the variable. Any ideas?

[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp]: `row` is already defined on the line above, so we have to be explicit. I'll go with `image_row` then

**Comment by LucasChollet** [comment]
Red CI is unrelated.

**Review by nico — APPROVED** [review]
LG!

File: Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp [hunk 4]
```diff
@@ -284,16 +284,16 @@ class TIFFLoadingContext {
             auto decoded_stream = make<BigEndianInputBitStream>(move(decoded_strip));
```
Comment (gmta) [inline Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp:309]: Typos in commit message: "store" -> "stores" (2x), "relatively to start" -> "relative to the start"

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp:309]: Ouch... thanks!

Comment (gmta) [reply Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp:309]: It still says "relatively" instead of "relative"

Comment (LucasChollet) [reply Userland/Libraries/LibGfx/ImageFormats/TIFFLoader.cpp:309]: There you go

**Pull request merged**